paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngReader.kt

```diff
@@ -0,0 +1,264 @@
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IEND
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IHDR
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_COLOR_TYPE_RGB
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_COLOR_TYPE_RGBA
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_SIG
+import okio.Buffer
+import okio.BufferedSource
+import okio.FileSystem
+import okio.Path.Companion.toPath
+import okio.buffer
+import java.awt.image.BufferedImage
+import java.awt.image.ColorModel
+import java.awt.image.DataBufferInt
+import java.awt.image.Raster
+import java.io.Closeable
+import java.io.File
+import java.lang.IllegalStateException
+import java.util.zip.CRC32
+import java.util.zip.Inflater
+
+/**
+ * Provides a simple sub-spec implementation for reading APNG files. Only handles
+ * chunks which are implemented by the writer.
+ */
+internal class ApngReader(file: File) : Closeable {
+  private val crcEngine = CRC32()
+  private val fileHandle = FileSystem.SYSTEM.openReadOnly(file.path.toPath())
+  private val buffer = fileHandle.source().buffer()
+
+  private lateinit var scratchSpace: IntArray
+
+  private var frameInfo: FrameInfo? = null
+  private var hasAlpha = false
+  private var sequenceNumber = -1
+
+  var frameCount = 1
+  var frameNumber = 0
+  var width = 0
+  var height = 0
+
+  init {
+    val header = buffer.readByteArray(PNG_SIG.size.toLong())
+    if (!header.contentEquals(PNG_SIG)) {
+      throw IllegalStateException("Input is not a valid PNG")
+    }
+
+    processMetadata()
+  }
+
+  fun getNextFrame(): BufferedImage? {
+    if (finished()) return null
+
+    if (frameNumber == 0) {
+      processTo(IDAT)
+    } else {
+      processTo(FDAT)
+    }
+
+    return getFrame()
+  }
+
+  fun getDelay(): Pair<Int, Int> {
+    return frameInfo?.let { it.delayNumerator.toInt() to it.delayDenominator.toInt() } ?: (1 to 0)
+  }
+
+  fun reset() {
+    frameNumber = 0
+    sequenceNumber = -1
+    frameInfo = null
+
+    fileHandle.reposition(buffer, PNG_SIG.size.toLong())
+    processMetadata()
+  }
+
+  private fun processMetadata() {
+    processTo(IHDR)
+
+    if (buffer.peekNextChunkId().contentEquals(ACTL.bytes)) {
+      processTo(ACTL)
+    }
+  }
+
+  fun finished() = buffer.peekNextChunkId().contentEquals(IEND.bytes)
+
+  override fun close() {
+    fileHandle.close()
+    buffer.close()
+  }
+
+  private fun processTo(header: Header) {
+    do {
+      val (currentHeader, data) = buffer.readChunk()
+      when (currentHeader) {
+        IHDR -> data.parseIHDR()
+        ACTL -> data.parseACTL()
+        FCTL -> data.parseFCTL()
+        IDAT -> data.parseIDAT()
+        FDAT -> data.parseFDAT()
+        IEND -> throw IllegalStateException("PNG ended while processing to $header")
+      }
+    } while (!header.bytes.contentEquals(currentHeader.bytes) && !finished())
+  }
+
+  private fun BufferedSource.parseIHDR() {
```

Comment:
I believe this uses the words `parse` and `read` interchangeably? Maybe `process` also? 

My preference is to stick to one, `read` if at all possible

(If there’s functions that do different things, they should have different prefixes)